### PR TITLE
feat: purge session

### PR DESCRIPTION
**Pull request by aneojgurhem**

### CODE REVIEW

File: Common/tests/TestBase/SessionTableTestBase.cs [hunk 1]
```diff
@@ -644,12 +646,17 @@ public async Task CancelPurgedSessionShouldThrow()
...
+                                      .ConfigureAwait(false);
+
       await SessionTable.PurgeSessionAsync(sessionId,
+                                           session.CreationDate,
                                            CancellationToken.None)
                         .ConfigureAwait(false);
 
-      var session = await SessionTable.GetSessionAsync(sessionId)
-                                      .ConfigureAwait(false);
+      session = await SessionTable.GetSessionAsync(sessionId)
```
[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Common/tests/TestBase/SessionTableTestBase.cs:658]: c'est plus haut dans le code...

**Review by ngruelaneo — APPROVED** [review]

File: Adaptors/Redis/src/ObjectStorage.cs [hunk 2]
```diff
@@ -145,16 +145,31 @@ await taskList.WhenAll()
...
+                           cancellationToken)
+        .ConfigureAwait(false);
+    }
+  }
+
+  /// <inheritdoc />
+  public IAsyncEnumerable<string> ListKeysAsync(CancellationToken cancellationToken = default)
+    => throw new NotImplementedException();
+
+  private async Task<bool> TryDeleteAsync(string            key,
```
Comment (lemaitre-aneo) [inline Adaptors/Redis/src/ObjectStorage.cs:163]:
```suggestion
  private async Task TryDeleteOneAsync(string            key,
```

File: Adaptors/S3/src/ObjectStorage.cs [hunk 3]
```diff
@@ -255,6 +237,37 @@ public IAsyncEnumerable<string> ListKeysAsync(CancellationToken cancellationToke
     return sizeBox[0];
   }
 
+  private async Task<bool> TryDeleteAsync(string            key,
```
Comment (lemaitre-aneo) [inline Adaptors/S3/src/ObjectStorage.cs:240]:
```suggestion
  private async Task TryDeleteOneAsync(string            key,
```

File: Adaptors/S3/src/ObjectStorage.cs [hunk 4]
```diff
@@ -255,6 +237,37 @@ public IAsyncEnumerable<string> ListKeysAsync(CancellationToken cancellationToke
...
+    {
+      var objectDeleteRequest = new DeleteObjectRequest
+                                {
+                                  BucketName = options_.BucketName,
+                                  Key        = objectStorageFullName,
+                                };
+      await s3Client_.DeleteObjectAsync(objectDeleteRequest,
+                                        cancellationToken)
+                     .ConfigureAwait(false);
+      logger_.LogInformation("Deleted data with {id}",
```
Comment (lemaitre-aneo) [inline Adaptors/S3/src/ObjectStorage.cs:257]:
```suggestion
      logger_.LogeDebug("Deleted data with {resultId}",
```

File: Common/src/Storage/IObjectStorage.cs [hunk 5]
```diff
@@ -61,14 +61,14 @@ public interface IObjectStorage : IInitializable
   /// <summary>
   ///   Delete data in the object storage
   /// </summary>
-  /// <param name="key">Key representing the object to delete</param>
+  /// <param name="keys">Keys representing the object to delete</param>
```
Comment (lemaitre-aneo) [inline Common/src/Storage/IObjectStorage.cs:64]:
```suggestion
  /// <param name="keys">Keys representing the objects to delete</param>
```

File: Common/src/Storage/ResultLifeCycleHelper.cs [hunk 6]
```diff
@@ -104,4 +106,29 @@ public static class ResultLifeCycleHelper
...
+  public static async Task PurgeResultsAsync(IResultTable      resultTable,
+                                             IObjectStorage    objectStorage,
+                                             string            sessionId,
+                                             CancellationToken cancellationToken)
+  {
+    await foreach (var ids in resultTable.GetResults(result => result.SessionId == sessionId,
+                                                     result => result.ResultId,
+                                                     cancellationToken)
+                                         .ToChunksAsync(500,
+                                                        TimeSpan.FromMilliseconds(100),
```
Comment (lemaitre-aneo) [inline Common/src/Storage/ResultLifeCycleHelper.cs:119]: I don't think you need a timeout

File: Common/src/gRPC/Services/GrpcSessionsService.cs [hunk 7]
```diff
@@ -265,24 +271,35 @@ public class GrpcSessionsService : Sessions.SessionsBase
...
+                                                        context.CancellationToken)
+                                       .ConfigureAwait(false);
+
+      await ResultLifeCycleHelper.PurgeResultsAsync(resultTable_,
+                                                    objectStorage_,
+                                                    request.SessionId,
+                                                    context.CancellationToken)
+                                 .ConfigureAwait(false);
+
       return new PurgeSessionResponse
```
Comment (lemaitre-aneo) [inline Common/src/gRPC/Services/GrpcSessionsService.cs:284]: You should have a logInfo here

File: Common/src/gRPC/Services/Submitter.cs [hunk 8]
```diff
@@ -518,16 +518,6 @@ await foreach (var taskRequest in taskRequests.WithCancellation(cancellationToke
...
-
-
-      //Discard value is used to remove warnings CS4014 !!
-      _ = Task.Factory.StartNew(async () => await objectStorage_.TryDeleteAsync(taskData.TaskId,
-                                                                                CancellationToken.None)
-                                                                .ConfigureAwait(false),
-                                cancellationToken);
-
-      logger_.LogInformation("Remove input payload of {task}",
-                             taskData.TaskId);
```
Comment (lemaitre-aneo) [inline Common/src/gRPC/Services/Submitter.cs:530]: You should keep this for now. It enables early deletion of payloads during the session processing.

File: Adaptors/S3/src/ObjectStorage.cs [hunk 9]
```diff
@@ -255,6 +237,34 @@ public IAsyncEnumerable<string> ListKeysAsync(CancellationToken cancellationToke
...
+                                  BucketName = options_.BucketName,
+                                  Key        = objectStorageFullName,
+                                };
+      await s3Client_.DeleteObjectAsync(objectDeleteRequest,
+                                        cancellationToken)
+                     .ConfigureAwait(false);
+      logger_.LogInformation("Deleted data with {resultId}",
+                             key);
+    }
+    catch (Exception ex)
```
[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Adaptors/S3/src/ObjectStorage.cs:260]: I don't know. I took what was there

[the remark this answers is not in the document]
Comment (aneojgurhem) [reply Adaptors/S3/src/ObjectStorage.cs]: It seems that the try catch is not needed

File: Adaptors/S3/src/ObjectStorage.cs [hunk 10]
```diff
@@ -156,34 +156,10 @@ public Task<HealthCheckResult> Check(HealthCheckTag tag)
...
-                       options_.BucketName);
-      return false;
-    }
-
-    return true;
-  }
+  public async Task TryDeleteAsync(IEnumerable<string> keys,
+                                   CancellationToken   cancellationToken = default)
+    => await keys.ParallelForEach(key => TryDeleteAsync(key,
+                                                        cancellationToken));
```
Comment (lemaitre-aneo) [inline Adaptors/S3/src/ObjectStorage.cs:162]: ConfigureAwait

**Pull request merged**